src/libos/crates/sgx-disk/src/host_disk/sync_io_disk.rs

```diff
@@ -104,6 +89,58 @@ impl SyncIoDisk {
         let end_offset = end_block * BLOCK_SIZE;
         Ok((begin_offset, end_offset))
     }
+
+    fn preadv(&self, fd: i32, bufs: &mut [BlockBuf], offset: i64) -> Result<usize> {
+        let iovecs: Box<Vec<libc::iovec>> = Box::new(
+            bufs.iter_mut()
+                .map(|buf| libc::iovec {
+                    iov_base: buf.as_slice_mut().as_mut_ptr() as _,
+                    iov_len: buf.as_slice().len(),
+                })
+                .collect(),
+        );
+        let (iovecs_ptr, iovecs_len) =
+            { (iovecs.as_ptr() as *const libc::iovec, iovecs.len() as i32) };
+
+        let read_len = unsafe {
+            #[cfg(not(feature = "sgx"))]
+            let read_len = libc::preadv64(fd, iovecs_ptr, iovecs_len, offset);
+            #[cfg(feature = "sgx")]
+            let read_len = libc::ocall::preadv64(fd, iovecs_ptr, iovecs_len, offset);
+            read_len
+        };
+        if read_len < 0 {
+            return_errno!(Errno::from(libc_errno() as u32), "libc::preadv64 error");
+        }
```

Comment:
Using the result from ocall without check is prone to Iago attacks.

After examining [the code of `libc::ocall::preadv64`](https://github.com/occlum/incubator-teaclave-sgx-sdk/blob/c2d700f8a622745204bf7d270fe85f2b9d77925e/sgx_libc/src/linux/x86_64/ocall.rs#L1845), I reach the conclusion that we do need to do the extra sanity check that `read_len` is not larger than the combined length of iovecs.

Reply:
Should we return `EINVAL` if `read_len`/`write_len` is not exactly equal to the combined length of all the iovecs?
Or we send an error log and return an `Ok` (which means allow partial preads/pwrites succeed)?

Reply:
To confirm, Partial reads/writes are unlikely to happen since buffers and offset are all within legal range of the host file (file size is truncated to size of total blocks when created)